Auto-complete metafields depending on shop data

aswamy opened the pull request:
## What are you adding in this PR?

Part 1 for https://github.com/Shopify/theme-tools/issues/502
- Adding support for metadata autocompletion on our editors
- listens to file changes for `.shopify` folder
- This should work for VScode only


## What's next? Any followup issues?

- We need to make it work on OSE next
- We will create a CLI command to fetch the metafields.json
- We need to automatically call this CLI command in VSCode on startup

## What did you learn?

- registering for `DidChangeWatchedFilesNotification` took some time, but had to read-up on client capabilities and how to subscribe to notifications after initialization

https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_didChangeWatchedFiles

## Tophat

Run the following query and dump the contents into `.shopify/metafields.json` in the theme repository. Remove the `nodes` array, and place it directory under each group.

```
query {
	article: metafieldDefinitions(ownerType: ARTICLE, first: 250) {
		nodes {
			...MetafieldDefinitionFields
		}
	}
	blog: metafieldDefinitions(ownerType: BLOG, first: 250) {
		nodes {
			...MetafieldDefinitionFields
		}
	}
	brand: metafieldDefinitions(ownerType: BRAND, first: 250) {
		nodes {
			...MetafieldDefinitionFields
		}
	}
	collection: metafieldDefinitions(ownerType: COLLECTION, first: 250) {
		nodes {
			...MetafieldDefinitionFields
		}
	}
	company: metafieldDefinitions(ownerType: COMPANY, first: 250) {
		nodes {
			...MetafieldDefinitionFields
		}
	}
	company_location: metafieldDefinitions(ownerType: COMPANY_LOCATION, first: 250) {
		nodes {
			...MetafieldDefinitionFields
		}
	}
	location: metafieldDefinitions(ownerType: LOCATION, first: 250) {
		nodes {
			...MetafieldDefinitionFields
		}
	}
	market: metafieldDefinitions(ownerType: MARKET, first: 250) {
		nodes {
			...MetafieldDefinitionFields
		}
	}
	order: metafieldDefinitions(ownerType: ORDER, first: 250) {
		nodes {
			...MetafieldDefinitionFields
		}
	}
	page: metafieldDefinitions(ownerType: PAGE, first: 250) {
		nodes {
			...MetafieldDefinitionFields
		}
	}
	product: metafieldDefinitions(ownerType: PRODUCT, first: 250) {
		nodes {
			...MetafieldDefinitionFields
		}
	}
	variant: metafieldDefinitions(ownerType: PRODUCTVARIANT, first: 250) {
		nodes {
			...MetafieldDefinitionFields
		}
	}
	shop: metafieldDefinitions(ownerType: SHOP, first: 250) {
		nodes {
			...MetafieldDefinitionFields
		}
	}
}

fragment MetafieldDefinitionFields on MetafieldDefinition {
	name
	namespace
	description
	type {
		category
		name
	}
}
```

## Before you deploy

- [x] I included a minor bump `changeset`
- [x] My feature is backward compatible

aswamy commented on packages/theme-check-common/src/index.ts:
```diff
@@ -74,6 +75,7 @@ export async function check(
     getDefaultTranslations: makeGetDefaultTranslations(fs, theme, rootUri),
     getDefaultSchemaLocale: makeGetDefaultSchemaLocale(fs, rootUri),
     getDefaultSchemaTranslations: makeGetDefaultSchemaTranslations(fs, theme, rootUri),
+    getMetafieldDefinitions: () => Promise.resolve({} as MetafieldDefinitionMap),
```
Theme check doesn't do anything yet, so we stubbed the method. But i can change it even though it doesn't get used.

charlespwd replied:
Bah might as well do it. Would be easy to forget :P

aswamy commented on packages/theme-check-common/src/index.ts:
```diff
@@ -74,6 +75,7 @@ export async function check(
     getDefaultTranslations: makeGetDefaultTranslations(fs, theme, rootUri),
     getDefaultSchemaLocale: makeGetDefaultSchemaLocale(fs, rootUri),
     getDefaultSchemaTranslations: makeGetDefaultSchemaTranslations(fs, theme, rootUri),
+    getMetafieldDefinitions: makeGetMetafieldDefinitions(fs),
```
Since `makeGetMetafieldDefinitions` reads the file on the filesystem, wouldn't this actually run every time we run check?

I feel like we need to memoize it too (like we did in startServer)

charlespwd replied:
hmm yeah. I think you'll need to do two things here:
- to support caching in language server, we inject the memoized version of the function as an optional dep
- to support CLI use cases, we do makeGetMetafieldDefinitions(fs) as a fallback

this means:
```ts
interface Dependencies {
  getMetafieldDefinitions?: () => Promise<MetafieldDefinition[]>
}

// ... 

const dependencies: AugmentedDependencies = {
   ...injectedDependencies,
   // ...
   getmetafieldDefinitions = injectedDependencies.getMetafieldDefinitions ?? makeGetMetafieldDefinitions(fs),
}
```

aswamy commented:
- updated to remove unused `getMetafieldDefinitions` in unit tests
- updated usage of `findThemeRootURI(...)` within `onDidChangeWatchedFiles`
- changed `getMetafieldDefinitions` inside theme-check so it can be injected

charlespwd commented on packages/theme-language-server-common/src/TypeSystem.ts:
```diff
@@ -122,8 +127,24 @@ export class TypeSystem {
+
+      if (!metafieldsProperty) continue;
+
+      metafieldsProperty.return_type = [{ type: `${category}_metafields`, name: '' }];
```
Hmm. Thinking about this, are we sure we're not mutating the original object here? Should we clone/replace metafieldsProperty before doing this mutation? 

Remember that `...objectMap` gives you a _shallow_ clone of the object map. Looks like we're mutating the original object map here and this is no bueno. Means that if you'd get metafield data from one theme in another.

aswamy replied:
We talked about it in person - since this isn't unique to a theme, we should be fine with this change happening across themes within a workspace. Added a warning message in a comment above the block instead.

charlespwd commented on packages/theme-language-server-common/src/TypeSystem.ts:
```diff
@@ -845,6 +931,101 @@ function settingReturnType(setting: InputSetting): ObjectEntry['return_type'] {
+    ...new Set([...Object.values(METAFIELD_TYPE_TO_TYPE), ...FETCHED_METAFIELD_CATEGORIES]),
+  ].reduce((map, type) => {
+    {
+      const metafieldEntry = JSON.parse(JSON.stringify(baseMetafieldEntry)); // easy deep clone
```
Oh are we doing it already. I'll just dm you to verify :D

aswamy replied:
Yeah sorry there are three things happening to make the whole thing work. We can talk about it more today, but here are the things:
- `customMetafieldTypeEntries` --> we copy the `metafield` type and create one for each category we support. e.g. `product_metafield`, `collection_metafield`, etc.
- `metafieldDefinitionsObjectMap` --> reads the metafield definitions we fetched and converts it into types for auto-completion
- the for-loop you commented above --> we override existing types like `product` and `collection`, so their `metafields` properties don't return as `untyped`, but instead return `product_metafields` and `collection_metafields` respectively

charlespwd reviewed (approved):
LGTM, didn't tophat yet :eyes:
